### PR TITLE
Prompt user on first push (containing a prefix) to store the path; combine defaultRegistry&defaultRegistryPath

### LINKED ISSUE #396: Fuse defaultRegistry and defaultRegistryPath settings into one setting.
Note: If a user already has both set, we should either warn the user about it, or change the setting ourselves.

### PR DESCRIPTION
**Pull request by PrashanthCorp**
Functionally important changes in the title

Chronologically - 
- Refactor RegisterCommand due to the max func length tslint rule - the function was at the limit of the number of lines. This prevents any change from happening there until this was addressed. 
- combine two workspace settings (defaultRegistry&defaultRegistryPath)
- Prompt user on first push (containing a prefix) to store the path

Fix #396 

Addresses #334

### CODE REVIEW

File: dockerExtension.ts [hunk 1]
```diff
@@ -166,13 +152,40 @@ export async function activate(ctx: vscode.ExtensionContext): Promise<void> {
...
+}
+
+export function deactivate(): Thenable<void> {
     if (!client) {
         return undefined;
     }
     // perform cleanup
     Configuration.dispose();
-    return await client.stop();
+    return client.stop();
```
Comment (StephenWeatherford) [inline dockerExtension.ts]: Why this change?

File: dockerExtension.ts [hunk 2]
```diff
@@ -166,13 +152,40 @@ export async function activate(ctx: vscode.ExtensionContext): Promise<void> {
...
+}
+
+function registerComposeCommands(): void {
+    registerCommand('vscode-docker.compose.up', composeUp);
+    registerCommand('vscode-docker.compose.down', composeDown);
+    registerCommand('vscode-docker.compose.restart', composeRestart);
+    registerCommand('vscode-docker.system.prune', systemPrune);
+}
+
+export function deactivate(): Thenable<void> {
```
Comment (StephenWeatherford) [inline dockerExtension.ts]: Why this change?

File: dockerExtension.ts [hunk 3]
```diff
@@ -166,13 +152,40 @@ export async function activate(ctx: vscode.ExtensionContext): Promise<void> {
     activateLanguageClient(ctx);
 }
 
-export async function deactivate(): Promise<void> {
+function registerContainerCommands(): void {
```
Comment (StephenWeatherford) [inline dockerExtension.ts]: Personally I'd rather not break them up into separate categories - it just becomes a nuisance.  And what if you add new categories, or a command doesn't fit neatly into one or the other?  I regret doing that in the storage extension.

What I would like to see:
1) Move vscode-docker.createWebApp's code into a function
2) move all of the registerCommands into a single separate function

File: dockerExtension.ts (hunk 1 above)
Comment (PrashanthCorp) [reply dockerExtension.ts]: The original function isn't async. I got an error regarding this. I'll admit this felt weird.

File: dockerExtension.ts (hunk 3 above)
Comment (PrashanthCorp) [reply dockerExtension.ts]: Sounds good. I see where that comes from. 

Thanks. Will get on it.

File: dockerExtension.ts (hunk 1 above)
Comment (StephenWeatherford) [reply dockerExtension.ts]: You mean deactivate?  This is what I see:

export async function deactivate(): Promise<void> {
    if (!client) {
        return undefined;
    }
    // perform cleanup
    Configuration.dispose();
    return await client.stop();
}

File: package.json [hunk 4]
```diff
@@ -242,12 +242,12 @@
         "docker.defaultRegistry": {
           "type": "string",
           "default": "",
-          "description": "Default registry when tagging an image, empty string will target Docker Hub when pushing."
+          "description": "(Deprecated) Default registry when tagging an image, empty string will target Docker Hub when pushing."
```
Comment (StephenWeatherford) [inline package.json]: I'd say we should go ahead and remove it so they get errors in the settings file.  You can still query for the value even if it's removed.

File: package.json [hunk 5]
```diff
@@ -242,12 +242,12 @@
...
         },
         "docker.defaultRegistryPath": {
           "type": "string",
           "default": "",
-          "description": "Path within registry to push to."
+          "description": "Default registry and path when tagging an image, empty string will target Docker Hub when pushing"
```
Comment (StephenWeatherford) [inline package.json]: empty string part no longer applies

File: README.md [hunk 6]
```diff
@@ -92,7 +92,7 @@ The Docker extension comes with a number of useful configuration settings allowi
 | `docker.attachShellCommand.windowsContainer` | Attach command to use for Windows containers | `powershell`
 | `docker.dockerComposeBuild` | Run docker-compose with the --build argument, defaults to true | `true`
 | `docker.dockerComposeDetached` | Run docker-compose with the --d (detached) argument, defaults to true | `true`
-| `docker.defaultRegistry` | Default registry when tagging an image, empty string will target Docker Hub when pushing. | `""`
+| `docker.defaultRegistry` | (Deprecated) Default registry when tagging an image, empty string will target Docker Hub when pushing. | `""`
```
Comment (StephenWeatherford) [inline README.md]: ditto (remove)

File: package.json (hunk 4 above)
Comment (PrashanthCorp) [reply package.json]: @fiveisprime your thoughts?

Comment (PrashanthCorp) [reply package.json]: I'll go ahead and remove this...

File: commands/tag-image.ts [hunk 7]
```diff
@@ -37,11 +37,6 @@ export async function tagImage(context?: ImageNode): Promise<void> {
             imageName = defaultRegistryPath + '/' + imageName;
         }
 
-        const defaultRegistry = configOptions.get('defaultRegistry', '');
```
Comment (StephenWeatherford) [inline commands/tag-image.ts]: Weren't you going to read the old setting, combine with the new one, and resave to the new one?

File: package.json (hunk 5 above)
Comment (PrashanthCorp) [reply package.json]: The docker command still assumes no prefix as pushing to Dockerhub. 
image

Does it make sense to expose Docker's assumption as part of this statement?

File: commands/tag-image.ts (hunk 7 above)
Comment (PrashanthCorp) [reply commands/tag-image.ts]: Combining happens as part of `consolidateDefaultRegistrySettings`, which I intend to run on extension activation. 

Note: There is a change in that function -having the config setting update be async- that hasn't been pushed. I'm pushing another commit that fixes that, but also implements part 1 of the bug filed (ask to save setting on first push).

File: package.json [hunk 8]
```diff
@@ -239,15 +239,15 @@
...
-          "default": "",
-          "description": "Default registry when tagging an image, empty string will target Docker Hub when pushing."
-        },
         "docker.defaultRegistryPath": {
           "type": "string",
           "default": "",
-          "description": "Path within registry to push to."
+          "description": "Default registry and path when tagging an image, empty string will target Docker Hub when pushing"
+        },
+        "docker.askToSavePrefix": {
```
Comment (StephenWeatherford) [inline package.json]: Why adding another concept of prefix here?  We already have the concept of defaultRegistryPath, so how about askToSaveRegistryPath?

File: commands/push-image.ts [hunk 9]
```diff
@@ -12,6 +13,10 @@ const teleCmdId: string = 'vscode-docker.image.push';
 const teleAzureId: string = 'vscode-docker.image.push.azureContainerRegistry';
 
 export async function pushImage(context?: ImageNode): Promise<void> {
+    const configOptions: vscode.WorkspaceConfiguration = vscode.workspace.getConfiguration('docker');
+    let askToSavePrefix = configOptions.get('askToSavePrefix', undefined);
```
Comment (StephenWeatherford) [inline commands/push-image.ts]: configOptions.get<boolean>(...

Comment (StephenWeatherford) [inline commands/push-image.ts]: default should be true not undefined

File: commands/push-image.ts [hunk 10]
```diff
@@ -26,6 +31,19 @@ export async function pushImage(context?: ImageNode): Promise<void> {
         }
     }
 
+    if (imageName.includes('/')) {
+        prefix = imageName.substring(0, imageName.lastIndexOf('/'));
+    }
+    if (prefix && askToSavePrefix !== false) { //account for undefined
+        let userPrefixPreference: vscode.MessageItem = await ext.ui.showWarningMessage("Would you like to save the prefix for autocomplete later?", DialogResponses.yes, DialogResponses.no, DialogResponses.skipForNow);
```
Comment (StephenWeatherford) [inline commands/push-image.ts]: `Would you like to save '${prefix}' as your default registry path?`

File: commands/push-image.ts [hunk 11]
```diff
@@ -26,6 +31,19 @@ export async function pushImage(context?: ImageNode): Promise<void> {
...
+        if (userPrefixPreference === DialogResponses.yes || userPrefixPreference === DialogResponses.no) {
+            askToSavePrefix = false;
+        }
+        if (userPrefixPreference === DialogResponses.yes) {
+            await configOptions.update('defaultRegistryPath', prefix, vscode.ConfigurationTarget.Workspace);
+        }
+        await configOptions.update('askToSavePrefix', askToSavePrefix, vscode.ConfigurationTarget.Workspace);
```
Comment (StephenWeatherford) [inline commands/push-image.ts]: nit: I'd just do 

await configOptions.update('askToSavePrefix', false, vscode.ConfigurationTarget.Workspace); 

instead of 'askToSavePrefix = false' above, rather than updating it even when you're not actually changing its value.

File: commands/push-image.ts [hunk 12]
```diff
@@ -26,6 +31,19 @@ export async function pushImage(context?: ImageNode): Promise<void> {
...
+    if (imageName.includes('/')) {
+        prefix = imageName.substring(0, imageName.lastIndexOf('/'));
+    }
+    if (prefix && askToSavePrefix !== false) { //account for undefined
+        let userPrefixPreference: vscode.MessageItem = await ext.ui.showWarningMessage("Would you like to save the prefix for autocomplete later?", DialogResponses.yes, DialogResponses.no, DialogResponses.skipForNow);
+        if (userPrefixPreference === DialogResponses.yes || userPrefixPreference === DialogResponses.no) {
+            askToSavePrefix = false;
+        }
+        if (userPrefixPreference === DialogResponses.yes) {
+            await configOptions.update('defaultRegistryPath', prefix, vscode.ConfigurationTarget.Workspace);
```
Comment (StephenWeatherford) [inline commands/push-image.ts]: After updating:

  showInformationMessage('You can change this value at any time via the docker.defaultRegistryPath setting.');

File: dockerExtension.ts [hunk 13]
```diff
@@ -157,13 +168,23 @@ export async function activate(ctx: vscode.ExtensionContext): Promise<void> {
...
-    activateLanguageClient(ctx);
+    if (defaultRegistry) {
+        if (defaultRegistryPath) { // combine both the settings, then notify the user
+            await configOptions.update('defaultRegistryPath', `${defaultRegistry}/${defaultRegistryPath}`, vscode.ConfigurationTarget.Workspace);
+        }
+        if (!defaultRegistryPath) {// assign defaultRegistry to defaultRegistryPath
+            await configOptions.update('defaultRegistryPath', `${defaultRegistry}`, vscode.ConfigurationTarget.Workspace);
+        }
+        await configOptions.update('defaultRegistry', '', false);
+        vscode.window.showInformationMessage("Moving forward, we shall deprecate the use of the defaultRegistry setting, combining it with defaultRegistryPath.")
```
Comment (StephenWeatherford) [inline dockerExtension.ts]: This implies you're *going* to do this, but you just did it.

`The 'docker.defaultRegistry' setting is now obsolete, and you should just use the 'docker.defaultRegistryPath' setting. Your settings have been updated to reflect this change.`

File: dockerExtension.ts [hunk 14]
```diff
@@ -157,13 +168,23 @@ export async function activate(ctx: vscode.ExtensionContext): Promise<void> {
...
 
-    activateLanguageClient(ctx);
+    if (defaultRegistry) {
+        if (defaultRegistryPath) { // combine both the settings, then notify the user
+            await configOptions.update('defaultRegistryPath', `${defaultRegistry}/${defaultRegistryPath}`, vscode.ConfigurationTarget.Workspace);
+        }
+        if (!defaultRegistryPath) {// assign defaultRegistry to defaultRegistryPath
+            await configOptions.update('defaultRegistryPath', `${defaultRegistry}`, vscode.ConfigurationTarget.Workspace);
+        }
+        await configOptions.update('defaultRegistry', '', false);
```
Comment (StephenWeatherford) [inline dockerExtension.ts]: I *think* you should leave off the false parameter.

File: package.json (hunk 5 above)
Comment (StephenWeatherford) [reply package.json]: Okay, I guess that makes sense.

File: commands/push-image.ts [hunk 15]
```diff
@@ -34,15 +34,16 @@ export async function pushImage(context?: ImageNode): Promise<void> {
...
+    if (prefix && askToSaveRegistryPath !== false) { //account for undefined
+        let userPrefixPreference: vscode.MessageItem = await ext.ui.showWarningMessage(`Would you like to save '${prefix}' as your default registry path?`, DialogResponses.yes, DialogResponses.no, DialogResponses.skipForNow);
         if (userPrefixPreference === DialogResponses.yes || userPrefixPreference === DialogResponses.no) {
-            askToSavePrefix = false;
+            askToSaveRegistryPath = false;
+            await configOptions.update('askToSaveRegistryPath', false, vscode.ConfigurationTarget.Workspace);
         }
         if (userPrefixPreference === DialogResponses.yes) {
             await configOptions.update('defaultRegistryPath', prefix, vscode.ConfigurationTarget.Workspace);
+            await vscode.window.showInformationMessage('You can change this value at any time via the docker.defaultRegistryPath setting.');
```
Comment (StephenWeatherford) [inline commands/push-image.ts]: Don't await... It will block the user from continuing.

Let's change message to:
'Default registry path saved. You can change this value at any time via the docker.defaultRegistryPath setting.'

File: commands/push-image.ts (hunk 11 above)
Comment (StephenWeatherford) [reply commands/push-image.ts]: And it's also high priority to avoid writing when not needed now that we know updating a setting could write it into a different place than it currently is.

File: dockerExtension.ts (hunk 14 above)
Comment (StephenWeatherford) [reply dockerExtension.ts]: Workspace is good for now.
